File: src/test/java/guru/drinkit/controller/RecipeControllerTest.java
package guru.drinkit.controller;

import guru.drinkit.common.Criteria;
import guru.drinkit.domain.Recipe;
import guru.drinkit.repository.RecipeRepository;
import guru.drinkit.security.Role;
import org.junit.Before;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import static guru.drinkit.controller.RecipeController.RESOURCE_NAME;
import static org.mockito.Mockito.when;
import static org.springframework.http.HttpMethod.DELETE;
import static org.springframework.http.HttpMethod.GET;
import static org.springframework.http.HttpMethod.POST;
import static org.springframework.http.HttpMethod.PUT;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

/**
 * @author pkolmykov
 */
@SuppressWarnings("SpringJavaAutowiringInspection")
public class RecipeControllerTest extends AbstractMockMvcTest {

    @Autowired
    private RecipeRepository recipeRepository;

    @Before
    public void setUp() throws Exception {
        when(recipeRepository.findOne(1)).thenReturn(new Recipe());

    }

    @Test
    public void getRecipeById() throws Exception {
        verifyAccess(GET, RESOURCE_NAME + "/1", status().isOk());
    }

    @Test
    public void createRecipe() throws Exception {
        verifyAccess(POST, RESOURCE_NAME, new Recipe(), status().isCreated(), Role.ADMIN);
    }

    @Test
    public void searchRecipes() throws Exception {
        verifyAccess(GET, RESOURCE_NAME + "?criteria=" + objectMapper.writeValueAsString(new Criteria()), status().isOk());//// TODO: 4/7/2016
    }

    @Test
    public void deleteRecipe() throws Exception {
        verifyAccess(DELETE, RESOURCE_NAME + "/1", status().isNoContent(), Role.ADMIN);
    }

    @Test
    public void updateRecipe() throws Exception {
        verifyAccess(PUT, RESOURCE_NAME + "/1", status().isNoContent(), Role.ADMIN);// TODO: 4/7/2016
    }

    @Test
    public void findRecipesByNamePart() throws Exception {
        verifyAccess(GET, RESOURCE_NAME + "/1", status().isOk());
    }

    @Test
    public void uploadMedia() throws Exception {
        verifyAccess(PUT, RESOURCE_NAME + "/1", status().isNoContent(), Role.ADMIN);/// TODO: 4/7/2016
    }

}

File: src/test/java/guru/drinkit/controller/RecipeControllerTest.kt
package guru.drinkit.controller

import guru.drinkit.controller.RecipeController.Companion.RESOURCE_NAME
import guru.drinkit.controller.common.AbstractMockMvcTest
import guru.drinkit.domain.Recipe
import guru.drinkit.repository.RecipeRepository
import guru.drinkit.security.Role
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.`when`
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status

/**
 * @author pkolmykov
 */
@Suppress("SpringKotlinAutowiredMembers")
class RecipeControllerTest : AbstractMockMvcTest() {

    @Autowired
    private lateinit var recipeRepository: RecipeRepository

    @Before
    fun setUp() {
        `when`(recipeRepository.findOne(1)).thenReturn(Recipe(name = ""))
        `when`(recipeRepository.save(anyObject<Recipe>())).thenReturn(Recipe(id = 1, name = ""))
    }

    @Test
    fun getRecipeById() {
        verifyAccess({ MockMvcRequestBuilders.get(RESOURCE_NAME + "/1") }, status().isOk)
    }

    @Test
    fun createRecipe() {
        val recipe = objectMapper.writeValueAsBytes(Recipe(ingredientsWithQuantities = listOf(Recipe.IngredientWithQuantity(1, 11, null)), name = ""))
        verifyAccess({ post(RESOURCE_NAME).content((recipe)).contentType(APPLICATION_JSON) },
                status().isCreated, Role.ADMIN)
    }

    @Test
    fun searchRecipes() {
        verifyAccess({
            MockMvcRequestBuilders.get(RESOURCE_NAME + "?criteria=null")
        }, status().isOk)
    }

    @Test
    fun deleteRecipe() {
        verifyAccess({ MockMvcRequestBuilders.delete(RESOURCE_NAME + "/1") },
                status().isNoContent, Role.ADMIN)
    }

    @Test
    fun updateRecipe() {
        val recipe = objectMapper.writeValueAsString(Recipe(id = 1, name = ""))
        verifyAccess({ put(RESOURCE_NAME + "/1").content(recipe).contentType(APPLICATION_JSON) }, status().isNoContent, Role.ADMIN)
    }

    @Test
    fun findRecipesByNamePart() {
        verifyAccess({
            MockMvcRequestBuilders.get(RESOURCE_NAME + "/1")
        }, status().isOk)
    }

    @Test
    fun uploadMedia() {
        verifyAccess({
            post(RESOURCE_NAME + "/1/media")
                    .content("{}")
                    .contentType(APPLICATION_JSON)
        }, status().isBadRequest, Role.ADMIN)
    }

}